File: 0894. All Possible Full Binary Trees/dp(top-down).java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<TreeNode> allPossibleFBT(int n) {
        
        Map<Integer, List<TreeNode>> memo = new HashMap<>();
        return build(n, memo);
        
    }
    
    public List<TreeNode> build(int n, Map<Integer, List<TreeNode>> memo){
        
        if(memo.containsKey(n)) return memo.get(n);
        
        List<TreeNode> res = new ArrayList<TreeNode>();

        if(n == 1){
            res.add(new TreeNode(0));
            return res;
        }
        
        for(int i = 1; i < n - 1; i++){
            if(i % 2 == 0) continue;
            int rightCount = n - i - 1;
            List<TreeNode> left = build(i, memo);
            List<TreeNode> right = build(rightCount, memo);
            for(TreeNode lNode: left){
                for(TreeNode rNode: right){
                    TreeNode root = new TreeNode(0, lNode, rNode);
                    res.add(root);
                }
            }
        }
        
        memo.put(n, res);
        return memo.get(n);
    }
}

File: 0894. All Possible Full Binary Trees/dp(top-down).js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
 var allPossibleFBT = function(n) {
    
    if(n % 2 === 0) return [];
    if(n === 0) return [];
    
    const dp = [...Array(n+1)].map(x => []);
    dp[1] = [new TreeNode(0)]
    
    for(let i = 3; i <= n; i++){
        if(i % 2 === 0) continue;
        for(let j = 1; j < i; j++){
            const leftSize = i - 1 - j;
            const rightSize = i - 1 - leftSize;
            for(const left of dp[leftSize]){
                for(const right of dp[rightSize]){
                    const root = new TreeNode(0)
                    root.left = left;
                    root.right = right;
                    dp[i].push(root)
                }
            }
            
        }
    }
    
    return dp[n]
};